core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -781,20 +782,7 @@ private[spark] class ExecutorAllocationManager(
           stageAttemptToNumRunningTask(stageAttempt) -= 1
           if (stageAttemptToNumRunningTask(stageAttempt) == 0) {
             stageAttemptToNumRunningTask -= stageAttempt
-            if (!stageAttemptToNumTasks.contains(stageAttempt)) {
-              val rpForStage = resourceProfileIdToStageAttempt.filter { case (k, v) =>
-                v.contains(stageAttempt)
-              }.keys
-              if (rpForStage.size == 1) {
-                // be careful about the removal from here due to late tasks, make sure stage is
-                // really complete and no tasks left
-                resourceProfileIdToStageAttempt(rpForStage.head) -= stageAttempt
-              } else {
-                logWarning(s"Should have exactly one resource profile for stage $stageAttempt," +
-                  s" but have $rpForStage")
-              }
-            }
-
+            removeStageFromResourceProfileIfUnused(stageAttempt)
```

Comment:
from jira:
> If a SparkListenerTaskEnd event for the last task in a stage was processed before SparkListenerStageCompleted for that stage,
then resourceProfileIdToStageAttempt will not be cleaned up properly.

I don't follow this, onTaskEnd is where resourceProfileIdToStageAttempt is cleaned up and removed, so how does  SparkListenerStageCompleted being called afterward affect it?  Is it being added back somehow that I"m not seeing?  If you can give the sequence of events that would be very helpful.

Reply:
oh wait, I think I see it.   stageAttemptToNumTasks only gets cleaned up in the onStageCompleted method.